src/stream.rs

```diff
@@ -411,326 +416,363 @@ mod tests {
+        assert_eq!(pop_len, 0);
+        assert_eq!(fin, false);
 
-        let read = buf.pop(std::usize::MAX).unwrap();
-        assert_eq!(read.len(), 19);
-        assert_eq!(read.fin(), true);
-        assert_eq!(&read[..], b"helloworldsomething");
-        assert_eq!(buf.len(), 0);
+        assert!(recv.push(first).is_ok());
+        assert_eq!(recv.len(), 19);
+        assert_eq!(recv.off(), 0);
 
-        let read = buf.pop(std::usize::MAX).unwrap();
-        assert_eq!(read.len(), 0);
+        let (pop_len, fin) = recv.pop(buf.as_mut_slice()).unwrap();
+        assert_eq!(pop_len, 19);
+        assert_eq!(fin, true);
+        assert_eq!(&buf[..pop_len], b"helloworldsomething");
+        assert_eq!(recv.len(), 0);
+        assert_eq!(recv.off(), 19);
+
+        let (pop_len, fin) = recv.pop(buf.as_mut_slice()).unwrap();
+        assert_eq!(pop_len, 0);
+        assert_eq!(fin, false);
     }
 
     #[test]
     fn split_read() {
-        let mut buf = RecvBuf::default();
-        assert_eq!(buf.len(), 0);
+        let mut recv = RecvBuf::default();
+        assert_eq!(recv.len(), 0);
+
+        let mut buf = vec![0; 32];
+        assert_eq!(buf.len(), 32);
 
         let first = RangeBuf::from(b"something", 0, false);
         let second = RangeBuf::from(b"helloworld", 9, true);
 
-        assert!(buf.push(first).is_ok());
-        assert_eq!(buf.len(), 9);
+        assert!(recv.push(first).is_ok());
+        assert_eq!(recv.len(), 9);
+        assert_eq!(recv.off(), 0);
 
-        assert!(buf.push(second).is_ok());
-        assert_eq!(buf.len(), 19);
+        assert!(recv.push(second).is_ok());
+        assert_eq!(recv.len(), 19);
+        assert_eq!(recv.off(), 0);
 
-        let read = buf.pop(10).unwrap();
-        assert_eq!(read.off(), 0);
-        assert_eq!(read.len(), 10);
-        assert_eq!(read.fin(), false);
-        assert_eq!(&read[..], b"somethingh");
-        assert_eq!(buf.len(), 9);
+        let (pop_len, fin) = recv.pop(&mut buf[..10]).unwrap();
+        assert_eq!(pop_len, 10);
+        assert_eq!(fin, false);
+        assert_eq!(&buf[..pop_len], b"somethingh");
+        assert_eq!(recv.len(), 9);
+        assert_eq!(recv.off(), 10);
 
-        let read = buf.pop(5).unwrap();
-        assert_eq!(read.off(), 10);
-        assert_eq!(read.len(), 5);
-        assert_eq!(read.fin(), false);
-        assert_eq!(&read[..], b"ellow");
-        assert_eq!(buf.len(), 4);
+        let (pop_len, fin) = recv.pop(&mut buf[..5]).unwrap();
+        assert_eq!(pop_len, 5);
+        assert_eq!(fin, false);
+        assert_eq!(&buf[..pop_len], b"ellow");
+        assert_eq!(recv.len(), 4);
+        assert_eq!(recv.off(), 15);
 
-        let read = buf.pop(10).unwrap();
-        assert_eq!(read.off(), 15);
-        assert_eq!(read.len(), 4);
-        assert_eq!(read.fin(), true);
-        assert_eq!(&read[..], b"orld");
-        assert_eq!(buf.len(), 0);
+        let (pop_len, fin) = recv.pop(&mut buf[..4]).unwrap();
```

Comment:
This should be `&mut buf[..10]` as in the previous test.

Reply:
I think it's ok, there should only be 4 bytes left in the stream on the last read?

Reply:
Well, the test was checking that trying to read more bytes than there are in the `RecvBuf` works, so that seems like a useful test to keep.